src/internal/analytics-metadata/metadata-utils.ts

```diff
@@ -0,0 +1,51 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { GeneratedAnalyticsMetadataFragment } from './interfaces';
+import { processLabel } from './labels-utils';
+
+export const mergeMetadata = (
+  metadata: GeneratedAnalyticsMetadataFragment | null,
+  localMetadata: GeneratedAnalyticsMetadataFragment | null
+) => {
+  const output = merge(metadata, localMetadata);
+  if (output.component && output.component.name) {
+    output.contexts = [...(output.contexts || []), { type: 'component', detail: output.component }];
+    delete output.component;
+  }
+  return output;
+};
+
+export const processMetadata = (node: HTMLElement | null, localMetadata: any): GeneratedAnalyticsMetadataFragment => {
+  return Object.keys(localMetadata).reduce((acc: any, key: string) => {
+    if (key.toLowerCase().match(/label$/)) {
+      acc[key] = processLabel(node, localMetadata[key]);
+    } else if (typeof localMetadata[key] !== 'string') {
+      acc[key] = processMetadata(node, localMetadata[key]);
+    } else {
+      acc[key] = localMetadata[key];
+    }
+    return acc;
+  }, {});
+};
+
+export const merge = (inputTarget: any, inputSource: any): any => {
```

Comment:
Would using lodash merge be simpler here? I'm not sure what kind of footprint their function would be though.

Reply:
I think it would, but I also didn't want to add dependencies for such a simple function

Reply:
Should this be able to handle array properties?

Reply:
No need. The only array property that is conceived is the `contexts` array, which is only present in the target. Adding a unit test to cover this use case